### DISCUSSION TITLE
[@types/react] Bug after update to version 18.2.61

### DISCUSSION THREAD

**Discussion by vimerzzz** [DefinitelyTyped/DefinitelyTyped#68840] [Issues with a @types package] [answered]
Before this version, my code works normally, but after update @types/react to v18.2.61. it shows up the error:
```
Type 'ForwardRefExoticComponent<Omit<BootstrapDateFieldProps, "ref"> & RefAttributes<HTMLDivElement>>' is not assignable to type 'ComponentType<BaseSingleInputFieldProps<Moment | null, Moment, FieldSection, any>> | undefined'.
  Type 'ForwardRefExoticComponent<Omit<BootstrapDateFieldProps, "ref"> & RefAttributes<HTMLDivElement>>' is not assignable to type 'FunctionComponent<BaseSingleInputFieldProps<Moment | null, Moment, FieldSection, any>>'.
    Types of property 'propTypes' are incompatible.
      Type 'WeakValidationMap<Omit<BootstrapDateFieldProps, "ref"> & RefAttributes<HTMLDivElement>> | undefined' is not assignable to type 'WeakValidationMap<BaseSingleInputFieldProps<Moment | null, Moment, FieldSection, any>> | undefined'.
        Type 'WeakValidationMap<Omit<BootstrapDateFieldProps, "ref"> & RefAttributes<HTMLDivElement>>' is not assignable to type 'WeakValidationMap<BaseSingleInputFieldProps<Moment | null, Moment, FieldSection, any>>'.
          Types of property 'ref' are incompatible.
            Type 'Validator<LegacyRef<HTMLDivElement> | undefined> | undefined' is not assignable to type 'Validator<Ref<HTMLDivElement> | undefined> | undefined'
```
at line:
```
slots={{ field: BootstrapDateField, ...props.slots }}
```

Below is my component if you're interested
```ts
interface BootstrapFieldProps extends BsPrefixRefForwardingComponent<"input", FormControlProps> {
    inputRef?: React.Ref<any>;
    InputProps?: {
        ref?: React.Ref<any>;
        endAdornment?: React.ReactNode;
        startAdornment?: React.ReactNode;
    };
    error?: boolean;
    focused?: boolean;
    ownerState?: any;
}

type BootstrapFieldComponent = ((
    props: BootstrapFieldProps & React.RefAttributes<HTMLDivElement>,
) => React.JSX.Element) & { propTypes?: any };

const BootstrapField = React.forwardRef(
    (props: BootstrapFieldProps, ref: React.Ref<HTMLDivElement>) => {
        const {
            inputRef,
            InputProps: { ref: containerRef, startAdornment, endAdornment } = {},
            error,
            focused,
            ownerState,
            ...other
        } = props;

        const handleRef = useForkRef(containerRef, ref);

        return (
            <div className="d-flex align-items-center" ref={handleRef}>
                {startAdornment}
                <Form.Control autoComplete="none" type="text" ref={inputRef} {...other}></Form.Control>
                {endAdornment}
            </div>
        );
    },
) as BootstrapFieldComponent;

interface BootstrapDateFieldProps extends UseDateFieldProps<Moment>,
    BaseSingleInputFieldProps<
        Moment | null,
        Moment,
        FieldSection,
        DateValidationError
    > { }

const BootstrapDateField = React.forwardRef(
    (props: BootstrapDateFieldProps, ref: React.Ref<HTMLDivElement>) => {
        const {
            inputRef: externalInputRef,
            slots,
            slotProps,
            ...textFieldProps
        } = props;

        const {
            onClear,
            clearable,
            ref: inputRef,
            ...fieldProps
        } = useDateField<Moment, typeof textFieldProps>({
            props: textFieldProps,
            inputRef: externalInputRef,
        });

        /* If you don"t need a clear button, you can skip the use of this hook */
        const { InputProps: ProcessedInputProps, fieldProps: processedFieldProps } =
            useClearableField<
                {},
                typeof textFieldProps.InputProps,
                DateFieldSlotsComponent,
                DateFieldSlotsComponentsProps<Moment>
            >({
                onClear,
                clearable,
                fieldProps,
                InputProps: fieldProps.InputProps,
                slots,
                slotProps,
            });
        return (
            <BootstrapField
                ref={ref}
                inputRef={inputRef}
                {...processedFieldProps as any}
                InputProps={ProcessedInputProps}
            />
        );
    },
);

const BootstrapDatePicker = React.forwardRef(
    (props: DatePickerProps<Moment>, ref: React.Ref<HTMLDivElement>) => {
        return (
            <DatePicker
                ref={ref}
                {...props}
                dayOfWeekFormatter={(day) => day.toUpperCase()}
                slots={{ field: BootstrapDateField, ...props.slots }}
                slotProps={{
                    actionBar: {
                        actions: ["clear", "today", "accept"]
                    }
                }}
            />
        );
    },
);
```

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
`React.RefAttributes` is for the props you can pass to a component not the props your component sees. The distinction is important in all current React versions because you can pass in string refs. However, your component will never see string refs but just `React.Ref<Instance>`.


Take this example:

```tsx
const Component = forwardRef(function Component(props, ref) {
  console.log(ref)
})

class App extends React.Component {
  render() {
    return <Component ref="component" />
  }
}
```

It's important that `Component` doesn't get a string but a ref object or callback ref.

Without knowing `BsPrefixRefForwardingComponent`, I can't know if there truly is a bug. I recommend trying to reduce the example as much as possible so that it fits into a [TypeScript playground](https://www.typescriptlang.org/play) and any unrelated code removed.

**Reply by vimerzzz** [reply]
sorry. you can find `BsPrefixRefForwardingComponent` at here [https://github.com/react-bootstrap/react-bootstrap/blob/8c623894c6d387c8288d5a3c3d5bcbe9ecf47dd7/src/helpers.ts](https://github.com/react-bootstrap/react-bootstrap/blob/8c623894c6d387c8288d5a3c3d5bcbe9ecf47dd7/src/helpers.ts#L24)

**Reply by vimerzzz** [reply]
@eps1lon i've checked again, my code still runs normally, but my vscode show me above error after updating to version 18.2.61.

More information, i've just migrated from CRA to vite yesterday, so it runs normally in vite with `npm start`, but not in CRA, and, not when build with `npm run build` with vite

**Reply by vimerzzz** [reply]
@eps1lon i've tried reproducing the problem at here [CodeSandbox](https://codesandbox.io/p/devbox/agitated-neumann-gdw5kz?file=%2Fsrc%2FApp.tsx%3A12%2C31)

**Reply by eps1lon** [reply]
This is a lot of code to debug the problem. Can you remove as much as possible until it no longer reproduces and inline all absolutely necessary lines of code?

**Reply by vimerzzz** [reply]
okay, wait me a sec

**Reply by vimerzzz** [reply]
@eps1lon pls check it again [CodeSandbox](https://codesandbox.io/p/devbox/agitated-neumann-gdw5kz?file=%2Fsrc%2FBootstrapDatePicker.tsx%3A13%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt7zhhiw00073j6e4kzwacz0%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt7zhhiv00023j6ey68m1svv%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt7zhhiw00043j6eb808shsx%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt7zhhiw00063j6eno70o7ma%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B49.08886185385551%252C50.91113814614449%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt7zhhiv00023j6ey68m1svv%2522%253A%257B%2522id%2522%253A%2522clt7zhhiv00023j6ey68m1svv%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fnode_modules%252Freact-refresh%252Fruntime.js%2522%252C%2522id%2522%253A%2522clt88py6901dx3j6e0z4qotux%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clt8b77z700023j6enklr136o%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A13%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A13%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FBootstrapDatePicker.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clt8b77z700023j6enklr136o%2522%257D%252C%2522clt7zhhiw00063j6eno70o7ma%2522%253A%257B%2522id%2522%253A%2522clt7zhhiw00063j6eno70o7ma%2522%252C%2522activeTabId%2522%253A%2522clt88236q006p3j6emw0qsq4m%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clt80wzyb01sy3j6eh96ks0v3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clt88236q006p3j6emw0qsq4m%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clt7zhhiw00043j6eb808shsx%2522%253A%257B%2522id%2522%253A%2522clt7zhhiw00043j6eb808shsx%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt7zwcah01733j6etnp635cu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clt7zyap100j5djip3bl490pp%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clt7zwcah01733j6etnp635cu%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Reply by vimerzzz** [reply, marked as answer]
@eps1lon The issue was fixed after upgrading mui/x-date-pickers to v7.0.0 and migrating some code based on docs. Well. i don't know why

**Reply by eps1lon** [reply]
Thank you for double checking. Closing since the original issue is resolved.
